### PR TITLE
Introducing stack-based iterators.

### PR DESCRIPTION
**Pull request by lemire**
In issue https://github.com/RoaringBitmap/roaring/issues/94, we raised the concern that when iterating through a bitmap, the iterator might allocate memory. This was solved by @Oppen.

User @anacrolix raised a secondary concern: can the iterators themselves be "on stack" (heap-free). Now iterators are tiny objects (~100 bytes) and you usually make extensive use of them, so it should not be much of a concern. 

This PR introduces one way to do it, on stack...

```Go
i := IntIterator{}
i.Initialize(bm)
for i.HasNext() {
   i.Next()
}
```

as opposed to

```Go
i := bm.Iterator()
for i.HasNext() {
   i.Next()
}
```

The new code is more error-prone. E.g., you may forget to call `i.Initialize(bm)`. However, if you are reusing the same iterator instance, you save on allocations... It may also be slightly faster.

```
> go test -bench BenchmarkIteratorAlloc -benchmem -run -
cpu: Intel(R) Core(TM) i7-7700K CPU @ 4.20GHz
BenchmarkIteratorAlloc/simple_iteration_with_alloc-8               17515             68413 ns/op             112 B/op          1 allocs/op
BenchmarkIteratorAlloc/simple_iteration-8                          19900             65896 ns/op               0 B/op          0 allocs/op
BenchmarkIteratorAlloc/reverse_iteration_with_alloc-8              16543             71236 ns/op             112 B/op          1 allocs/op
BenchmarkIteratorAlloc/reverse_iteration-8                         20451             58762 ns/op               0 B/op          0 allocs/op
```

Note that this PR does not port the same idea to 64-bit roaring bitmaps. If we merge this PR, then I would encourage people to provide the same functionality to 64-bit roaring bitmaps.

### CODE REVIEW

**Comment by anacrolix** [comment]
Looks great. Note that this isn't technically stack-based, what it is is "recyclable" or "reusable" iterators. I suspect even in the benchmarks, the reused iterators are still being heap allocated, it's only allocating the once though. That's fine. If this is correct I would reword it to be reusable or recycleable, or caller-managed allocation.

**Comment by anacrolix** [comment]
My implementation from https://github.com/RoaringBitmap/roaring/issues/94#issuecomment-1122102645 uses the wording "recycleable". Also I think that type aliasing to a private type like I did might still mess with the documentation. I can't get pkg.go.dev to confirm for this commit.

**Comment by anacrolix** [comment]
Yeah all it gives is
```
type IntIterator = intIterator
    IntIterator is meant to allow you to iterate through the values of a bitmap,
    see Initialize(a *Bitmap)
```
Which doesn't show that you can use all the methods of intIterator I think.

**Comment by lemire** [comment]
I will hold on to this PR and not merge it.

If it helps you @anacrolix  and you can test it out, then we can merge it.

**Comment by anacrolix** [comment]
I am testing this, specifically at https://github.com/anacrolix/torrent/commit/cff7c0b9501a51236acbcef4f0f51f9b39294840#diff-bffc2681ca171fb12aa1c57dce92db8710256de0bbe848069c3dff825eee9299R199 and https://github.com/anacrolix/torrent/commit/cff7c0b9501a51236acbcef4f0f51f9b39294840#diff-449c382592420ab0e94eb0707f863e9fdaeceea483a1f39ebce93c42b57ba04dR2506.

**Comment by lemire** [comment]
@anacrolix Feel free to provide feedback.

If it helps you, we will check it in.

File: roaring.go [hunk 1]
```diff
@@ -555,19 +569,25 @@ func (rb *Bitmap) Iterate(cb func(x uint32) bool) {
 // Iterator creates a new IntPeekable to iterate over the integers contained in the bitmap, in sorted order;
 // the iterator becomes invalid if the bitmap is modified (e.g., with Add or Remove).
 func (rb *Bitmap) Iterator() IntPeekable {
-	return newIntIterator(rb)
+    p := new(intIterator)
```
Comment (Oppen) [inline roaring.go:572]: There seems to be a slight indentation issue here.

**Comment by Oppen** [comment]
> Note that this isn't technically stack-based, what it is is "recyclable" or "reusable" iterators.

As the type is exposed, you can create the instance in the stack. Value types from which no reference is created never escape to the heap in Go AFAIK.
Of course, the `New...` methods still escape, but that's desired as we shouldn't break the API unless there's really a big advantage from doing so, while there would be none compared to just creating the value if we intend on recycling, as we'll be calling initialize explicitly often anyway, it doesn't hurt to do it at creation time.


@lemire this looks real good. I don't know why I didn't think of implementing it this way back then.

**Comment by anacrolix** [comment]
It's running great in testing. It doesn't force you to heap-allocate, unlike the existing methods. However that's not where the benefit is coming from in this PR. IntIterator.Initialize still escapes to the heap due to the complexity of the method but it doesn't matter: The caller controls the once-off allocation cost. By having the caller provide a pointer to the value used for the iterator, the caller is controlling and enforcing reuse of the value. Wording-wise this is a reusable iterator. Recycle typically means that there's some action required to make it reusable, in this case there is not.

**Comment by lemire** [comment]
Let us merge this.

**Comment by jotto** [comment]
Thank you for this great library and this new feature.

Any ideas on a fast way to re-initialize an `IntIterator` to a previous state?

My attempt was `i2 := *i` recurse, and then `*i = i2`, but this seems like there should be a better/faster way.

```golang
func recurser(location uint32, i *roaring.IntIterator) {
        // fake code, not actually functional
	bm := getBitmap(location)
	i.Initialize(bm)
	for i.HasNext() {
		n := i.Next()
		
		i2 := *i
		recurser(n, i)
		*i = i2
	}
}
```

**Comment by lemire** [comment]
>  seems like there should be a better/faster way.

Can you elaborate?

**Comment by jotto** [comment]
Copying the pointer like I did in the example didn't seem to cost too much in my tests, but also seemed unnecessary if the internal fields of `IntIterator` were simple enough to copy instead (but a cursory glance makes the internals look too complicated to copy). Just curious if it seemed feasible off the top of your head.

My idea was something like this:

```golang
state := i.GetState()
recurse()
i.restoreState(state)
```
